README.md
# jasmine-console-reporter

[![npm](http://img.shields.io/npm/v/jasmine-console-reporter.svg?style=flat-square)](https://www.npmjs.com/package/jasmine-console-reporter)
[![release](https://img.shields.io/github/release/onury/jasmine-console-reporter.svg?style=flat-square)](https://github.com/onury/jasmine-console-reporter)
[![license](http://img.shields.io/npm/l/jasmine-console-reporter.svg?style=flat-square)](https://github.com/onury/jasmine-console-reporter/blob/master/LICENSE)
[![downloads](http://img.shields.io/npm/dm/jasmine-console-reporter.svg?style=flat-square)](https://www.npmjs.com/package/jasmine-console-reporter)
[![dependencies](https://david-dm.org/onury/jasmine-console-reporter.svg?style=flat-square)](https://david-dm.org/onury/jasmine-console-reporter)
[![maintained](https://img.shields.io/maintenance/yes/2018.svg?style=flat-square)](https://github.com/onury/jasmine-console-reporter/graphs/commit-activity)  

> © 2018, Onur Yıldırım ([@onury](https://github.com/onury)). MIT License.

Progressive Console Reporter for Jasmine. (for Jasmine v3+)  
Outputs detailed test results to the console, with beautiful layout and colors.  

Example screen:

<p align="center">
    <a href="https://github.com/onury/jasmine-console-reporter">
        <img src="https://raw.github.com/onury/jasmine-console-reporter/master/screenshots/jcr-3-screen.jpg" width="100%" alt="JCR Screen" />
    </a>
</p>

Overall summary outputs:

<p align="center">
    <a href="https://github.com/onury/jasmine-console-reporter">
        <img src="https://raw.github.com/onury/jasmine-console-reporter/master/screenshots/jcr-summary.jpg" width="100%" alt="JCR Summaries" />
    </a>
</p>

## Installation

```shell
npm i jasmine-console-reporter --save-dev
```
_JCR v3+ requires peer dependency Jasmine v3+ and works on Node.js v6+. See [compatibility table](#compatibility-table)._

## Usage

```js
const JasmineConsoleReporter = require('jasmine-console-reporter');
const reporter = new JasmineConsoleReporter({
    colors: 1,           // (0|false)|(1|true)|2
    cleanStack: 1,       // (0|false)|(1|true)|2|3
    verbosity: 4,        // (0|false)|1|2|(3|true)|4|Object
    listStyle: 'indent', // "flat"|"indent"
    timeUnit: 'ms',      // "ms"|"ns"|"s"
    timeThreshold: { ok: 500, warn: 1000, ouch: 3000 }, // Object|Number
    activity: false,     // boolean or string ("dots"|"star"|"flip"|"bouncingBar"|...)
    emoji: true,
    beep: true
});
// pass the initialized reporter to whichever task or host...
```

### Options

| Option   | Description |
| -------- | ----------- |
| <h4>**`colors`**</h4>`Number\|Boolean` | Default: `1`. Specifies whether the output should have colored text. Possible integer values: 0 to 2. Set to `1` (or `true`) to enable colors. Set to `2` to use the [ANSI escape codes](https://www.npmjs.com/package/chalk#chalkstyles). Option `2` can be useful if, for example, you're running your tests from a sub-process, and the colors aren't showing up.            |
| <h4>**`cleanStack`**</h4>`Number\|Boolean` | Default: `1`. Specifies the filter level for the error stacks. Possible integer values: 0 to 3. Set to `1` (or `true`) to only filter out lines with jasmine-core path from stacks. Set to `2` to filter out all `node_modules` paths. Set to `3` to also filter out lines with no file path in it.            |
| <h4>**`verbosity`**</h4>`Number\|Boolean\|Object` | Default: `4`. Specifies the verbosity level for the reporter output. Possible integer values: 0 to 4. When a `Boolean` value is passed, `true` defaults to `4` and `false` defaults to `0`. Level 0: reports errors only. Level 1: also displays a summary. Level 2: also reports pending specs. Level 3: additionally displays all suites and specs as a list, except disabled and excluded specs. Level 4: lists all. Or set to an object to toggle each output section individually. Any omitted value will default to `true`. e.g. `{ pending: false, disabled: false, specs: false, summary: true }` |
| <h4>**`listStyle`**</h4>`String` | Default: `"indent"`. Indicates the style of suites/specs list output. Possible values: `"flat"` or `"indent"`. Setting this to `"indent"` provides a better view especially when using nested (describe) suites. This option is only effective when verbosity level is set to `3`, `4` or `true`.            |
| <h4>**`timeUnit`**</h4>`String` | Default: `"ms"`. Specifies the time unit to be used to measure spec execution. `"ms"` for milliseconds, `"ns"` for milliseconds including hi-res nanoseconds remainder, `"s"` for seconds including milliseconds remainder. _Note that overall time in final summary is always reported in seconds._|
| <h4>**`timeThreshold`**</h4>`Number\|Object` | Default: `{ ok: 500, warn: 1000, ouch: 3000 }`. Specifies the expected maximum time(s) (in milliseconds) for any spec to complete its execution. If threshold is exceeded, elapsed time for that spec will be accented in the output. This is only effective if `colors` option is enabled. |
| <h4>**`activity`**</h4>`Boolean\|String` | Default: `false`. Specifies whether to enable the activity indicator animation that outputs the current spec that is being executed. Set to a string value to set/change the [spinner style](https://github.com/sindresorhus/cli-spinners/blob/master/spinners.json).|
| <h4>**`emoji`**</h4>`Boolean\|Object` | Default: `false`. Whether to output some emojis within the report, for a little fun. To customize emojis, you can set an object. _Note that emojis will be auto-disabled in CI environments._ |
| <h4>**`beep`**</h4>`Boolean` | Default: `true`. Whether to play system beep when tests finish with status "failed". _Note that beep will be auto-disabled in CI and non-TTY environments._ |

### Full Example with Jasmine (Node.js)

```js
#!/usr/bin/env node

// setup Jasmine
const Jasmine = require('jasmine');
const jasmine = new Jasmine();
jasmine.loadConfig({
    spec_dir: 'test',
    spec_files: ['**/*[sS]pec.js'],
    helpers: ['helpers/**/*.js'],
    random: false,
    seed: null,
    stopSpecOnExpectationFailure: false
});
jasmine.jasmine.DEFAULT_TIMEOUT_INTERVAL = 15000;

// setup console reporter
const JasmineConsoleReporter = require('jasmine-console-reporter');
const reporter = new JasmineConsoleReporter({
    colors: 1,           // (0|false)|(1|true)|2
    cleanStack: 1,       // (0|false)|(1|true)|2|3
    verbosity: 4,        // (0|false)|1|2|(3|true)|4|Object
    listStyle: 'indent', // "flat"|"indent"
    timeUnit: 'ms',      // "ms"|"ns"|"s"
    timeThreshold: { ok: 500, warn: 1000, ouch: 3000 }, // Object|Number
    activity: true,
    emoji: true,         // boolean or emoji-map object
    beep: true
});

// initialize and execute
jasmine.env.clearReporters();
jasmine.addReporter(reporter);
jasmine.execute();
```

## Change-Log

See [CHANGELOG.md][changelog].

### Compatibility Table

| Jasmine Console Reporter | Node           | Jasmine   |
| ------------------------ | -------------- | --------- | 
| v3.x                     | v6 and newer   | v3.x      |
| v2.x                     | v4 and newer   | v2.x      |
| v1.2.7                   | below v4       | v2.x      |

## License

[MIT][license].


[changelog]:https://github.com/onury/jasmine-console-reporter/blob/master/CHANGELOG.md
[license]:https://github.com/onury/jasmine-console-reporter/blob/master/LICENSE
[grunt-jn]:https://github.com/onury/grunt-jasmine-nodejs


CHANGELOG.md
# Jasmine Console Reporter Changelog

All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org).

## [3.1.0](https://github.com/onury/jasmine-console-reporter/compare/v3.0.2...v3.1.0) (2018-08-21)

### Added
- Execution elapsed time for each spec. [#11](https://github.com/onury/jasmine-console-reporter/issues/11)
- New option: `timeUnit`  (`String`, default: `"ms"`) The time unit to be used to measure spec execution. `"ms"` for milliseconds, `"ns"` for milliseconds including hi-res nanoseconds remainder, `"s"` for seconds including milliseconds remainder.
- New option: `timeThreshold`  (`Object|Number`, default: `{ ok: 500, warn: 1000, ouch: 3000 }`) the expected maximum time(s) (in milliseconds) for any spec to complete its execution. If threshold is exceeded, elapsed time for that spec will be accented in the output. This is only effective if `colors` option is enabled.
- `verbosity` option now also accepts an object, in order to toggle each output section individually. [#11](https://github.com/onury/jasmine-console-reporter/issues/11)

### Changed
- Updated dependencies to latest versions.

### Fixed
- An issue where some options (such as `colors`) would not pick up default values when omitted. Fixes [#10](https://github.com/onury/jasmine-console-reporter/issues/10).


## [3.0.2](https://github.com/onury/jasmine-console-reporter/compare/v3.0.0...v3.0.2) (2018-05-17)

### Changed
- Randomness and random seed displayed if running in random order and `verbosity` is `1` or above. [PR #9](https://github.com/onury/jasmine-console-reporter/pull/9) by [@steverice](https://github.com/steverice).
- Updated dependencies to latest versions.


## [3.0.0](https://github.com/onury/jasmine-console-reporter/compare/v2.0.1...v3.0.0) (2018-04-03)
> _This major version is re-written in ES2015. See breaking changes below._

### Added
- Support for Jasmine v3+ (peer-dependency). Also see [Jasmine v3 changes](https://github.com/jasmine/jasmine/blob/v3.0.0/release_notes/3.0.md).
- Additional information is displayed, if running in *random* order and `verbosity` is set to `4`.
- Overall status (and reason if status is `incomplete`) within report summary.
- Support for non-interactive output terminals. (This allows piping output to a file.)
- New option: `emoji` 👊 (`boolean`, default: `false`). _Note: emojis will be auto-disabled in CI environments._
- New option: `beep` (`boolean`, default: `false`). _Note: beep will be auto-disabled in CI and non-TTY environments._
- Activity spinner styles. Instead of setting to `true`, set `activity` option to a string value to set the style of the spinner. e.g. `"dots"` (default) or `"bouncingBar"`. See all possible values [here](https://github.com/sindresorhus/cli-spinners/blob/master/spinners.json).

### Changed
- **BREAKING**: Requires Node.js v6 or newer.
- Improved exception stack parsing.
- Better activity spinner.

### Fixed
- An issue where exception message is not printed before the stack. Fixes [#7](https://github.com/onury/jasmine-console-reporter/issues/7) & [#8](https://github.com/onury/jasmine-console-reporter/issues/8).


## [2.0.1](https://github.com/onury/jasmine-console-reporter/compare/v1.2.8...v2.0.1) (2017-08-20)

### Changed
- **Breaking**: Due to [`chalk` dependency](https://github.com/chalk/chalk/releases/tag/v2.0.1) upgrade, the reporter now requires Node.js v4 or newer.
- Extensive code revision.

### Fixed
- An issue where the reporter would crash when `colors` option is set to `2` (ANSI colors). Fixes [issue #5](https://github.com/onury/jasmine-console-reporter/issues/5).


## [1.2.8](https://github.com/onury/jasmine-console-reporter/compare/v1.2.7...v1.2.8) (2017-08-20)

> _(**CAUTION**: version 1.2.8 has a [critical bug](https://github.com/onury/jasmine-console-reporter/issues/5). To avoid it, do NOT set `colors` option to `2` (ANSI colors). Or better update to v2+ if you're using Node.js version 4 or newer.)_

### Changed
- Updated `jasmine` and `chalk` to latest versions.
- Clean-up.


## [1.2.7](https://github.com/onury/jasmine-console-reporter/compare/v1.2.6...v1.2.7) (2016-08-21)

### Fixed
- An issue with `cleanStack` option, that would render the first line as the error message, not respecting messages with `\n` (new-line) in them.

### Changed
- Updated dev-dependencies.


## [1.2.6](https://github.com/onury/jasmine-console-reporter/compare/v1.2.4...v1.2.6) (2016-05-09)

### Fixed
- An issue with failed specs that don't have an error stack, would not output a warning message.

### Changed
- Updated dependencies to their latest versions.


## [1.2.4](https://github.com/onury/jasmine-console-reporter/compare/v1.2.2...v1.2.4) (2016-03-18)

### Added
- Support for ANSI escape codes, for `colors` option. (PR [#1](https://github.com/onury/jasmine-console-reporter/pull/1) by [@msrose](https://github.com/msrose))


## [1.2.2](https://github.com/onury/jasmine-console-reporter/compare/v1.2.0...v1.2.2) (2016-02-27)

### Added
- Grunt, jasmine tests, etc...

### Changed
- Moved helper classes and utils to separate modules.


## [1.2.0](https://github.com/onury/jasmine-console-reporter/compare/v1.1.3...v1.2.0) (2016-02-26)

### Changed
- Moved the reporter to its own repo.
- Code revisions and clean-up.


## [1.1.3](https://github.com/onury/jasmine-console-reporter/compare/v1.1.0...v1.1.3) (2015-07-05)

### Changed
- Expanded `verbosity` levels (0 to 4). Setting to `3` will not report disabled specs anymore while listing others. Set to `4` (default) for the most verbose report.
- Updated dependencies to their latest versions.


## [1.1.0](https://github.com/onury/jasmine-console-reporter/compare/v1.0.1...v1.1.0) (2015-05-01)

### Changed
- Updated dependencies.
- Minor code revisions and clean-up.



test/calc.js
module.exports = (() => {
    'use strict';

    function Calculator() {}

    Calculator.prototype.add = (a, b) => {
        return a + b;
    };
    Calculator.prototype.subtract = (a, b) => {
        return a - b;
    };
    Calculator.prototype.divide = (a, b) => {
        return a / b;
    };
    Calculator.prototype.multiply = (a, b) => {
        return a * b;
    };

    return Calculator;

})();


test/index.js
#!/usr/bin/env node

const Jasmine = require('jasmine');
const JasmineConsoleReporter = require('../index.js');
const jasmineConfig = require('./config/jasmine.config.json');

// setup Jasmine
const jasmine = new Jasmine();
jasmine.loadConfig(jasmineConfig);
jasmine.jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000;

console.log('');
console.log('--------------------------------------------------');

// setup reporter
const reporterConfigFile = process.argv[2] || './config/reporter.config-1.json';
console.log('DEVELOPMENT TEST:', reporterConfigFile);
console.log('--------------------------------------------------');
const reporterConfig = require(reporterConfigFile);
console.log('original config:', JSON.stringify(reporterConfig));
const reporter = new JasmineConsoleReporter(reporterConfig);
jasmine.env.clearReporters();
jasmine.addReporter(reporter);

// run
console.log('--------------------------------------------------');
console.log('config:', JSON.stringify(reporter.options));
console.log('--------------------------------------------------');
console.log('');
jasmine.execute();


test/calc.spec.js

const Calculator = require('./calc');

(function () {
    'use strict';

    describe('Calculator Suite', function () {
        var calculator,
            a = 10,
            b = 2;

        beforeAll(function () {
            calculator = new Calculator();
        });

        it('should add numbers', function () {
            var result = calculator.add(a, b);
            expect(result).toEqual(11);
            // this is a custom matcher implemented within our helper.
            expect(result).toBePositive();
        });
        it('should execute async spec...', function (done) {
            var result = calculator.add(a, b);
            setTimeout(function () {
                expect(result).toEqual(12);
                done();
            }, 2000);
        });
        it('should subtract numbers', function () {
            var result = calculator.subtract(a, b);
            expect(result).toEqual(8);
            // this is another custom matcher implemented within our helper.
            expect(result).toBePositive();
            pending();
        });
        // it('should divide numbers', function () {
        //     var result = calculator.divide(a, b);
        //     expect(result).toEqual(0); // should be 5 so this will throw
        // });
        it('should multiply numbers', function () {
            var result = calculator.multiply(a, b);
            expect(result).toEqual(20);
            pending('this is the pending reason');
        });

    });

})();


index.js
'use strict';

// dep modules
const ci = require('ci-info');

// own modules
const StreamPrinter = require('./lib/StreamPrinter');
const StyleFactory = require('./lib/StyleFactory');
const TestStats = require('./lib/TestStats');
const utils = require('./lib/utils');

const INDENT_CHAR = ' ';
const INDENT_UNIT = 3;

/**
 *  Jasmine Console Reporter
 *  @author   Onur Yıldırım <onur@cutepilot.com>
 *  @license  MIT
 */
class JasmineConsoleReporter {

    constructor(options = {}) {
        this.options = utils.getOptions(options);

        // disable emojis & beep on CI platforms
        if (ci.isCI) {
            this.options.emoji = false;
            this.options.beep = false;
        }

        const pOpts = typeof options.activity === 'string'
            ? { spinner: options.activity }
            : undefined;
        this.print = new StreamPrinter(pOpts);
        this.style = StyleFactory.create(this.options);
        this.stats = new TestStats(this.options);

        this.verbosity = this.options.verbosity;

        // Keeping track of suite (describe) nest levels.
        this._depth = -1;
        // Just keeping a flag to determine whether an extra new line is
        // needed when there is a remaining spec after a nested suite is
        // finished.
        this._isSuiteDone = false;
    }

    _specFailureDetails(spec, num) {
        this.print.line(this.style.red(num + '. '));
        const title = spec.fullName.replace(spec.description, ': ' + spec.description);
        this.print.str(this.style.cyan(title));
        let i, failedExpectation, errInfo;
        for (i = 0; i < spec.failedExpectations.length; i++) {
            failedExpectation = spec.failedExpectations[i];
            if (failedExpectation.stack) {
                errInfo = utils.restack(failedExpectation, this.options.cleanStack, this.style);
            } else {
                errInfo = 'Error: ' + (failedExpectation.message || 'Unknown Error');
                errInfo = this.style.red(errInfo);
            }
            this.print.line(utils.indent(errInfo, INDENT_UNIT));
        }
        this.print.newLine();
    }

    _specPendingDetails(spec, num) {
        this.print.line(this.style.yellow(num + '. '));
        const title = spec.fullName.replace(spec.description, ': ' + spec.description);
        this.print.str(this.style.cyan(title));
        const pendingReason = spec.pendingReason
            ? this.style.yellow('Reason: ' + spec.pendingReason)
            : this.style.gray('(No pending reason)');
        this.print.line(utils.indent(pendingReason, INDENT_UNIT));
        this.print.newLine();
    }

    _suiteFailureDetails(suite) {
        suite.failedExpectations.forEach(ex => {
            this.print.line(this.style.red('>> An error was thrown in an afterAll'));
            const stack = utils.restack(ex, this.options.cleanStack, this.style);
            this.print.line(utils.indent(stack, INDENT_UNIT));
        });
        this.print.newLine(2);
    }

    _finalReport(doneSummary) {
        if (this.stats.failedSpecList.length > 0) {
            this.print.line(this.style.white(this.style.underline('Failed Specs')) + this.style.white(':'));
            this.print.newLine();
            this.stats.failedSpecList.forEach((spec, index) => {
                this._specFailureDetails(spec, index + 1);
            });
        }

        if (this.verbosity.pending && this.stats.pendingSpecList.length > 0) {
            this.print.line(this.style.white(this.style.underline('Pending Specs')) + this.style.white(':'));
            this.print.newLine();
            this.stats.pendingSpecList.forEach((spec, index) => {
                this._specPendingDetails(spec, index + 1);
            });
        }

        if (this.verbosity.summary) {
            this.print.line(this.style.white(this.style.underline('Summary') + ':'));
            this.print.newLine();

            if (doneSummary && doneSummary.overallStatus) {
                const status = doneSummary.overallStatus;
                const oStyle = this.style.fromKeyword(status);
                this.print.line(this.style.emoji[status + 'Status'] + oStyle(utils.capitalize(status)));
                if (doneSummary.incompleteReason) {
                    this.print.str(this.style.gray(' (' + doneSummary.incompleteReason + ')'));
                }
            }

            if (this.stats.specs.total > 0) {
                const executedSuites = this.stats.suites.total - this.stats.suites.disabled;
                this.print.line('Suites:  ' + this.style.white(executedSuites) + ' of ' + this.stats.suites.total);
                if (this.stats.suites.disabled) {
                    this.print.str(this.style.yellow(' (' + this.stats.suites.disabled + ' disabled)'));
                }

                this.print.line('Specs:   ' + this.style.white(this.stats.executedSpecs) + ' of ' + this.stats.specs.defined);
                const specsInfo = [];
                if (this.stats.specs.pending) {
                    specsInfo.push(this.stats.specs.pending + ' pending');
                }

                if (this.stats.specs.disabled > 0) {
                    specsInfo.push(this.stats.specs.disabled + ' disabled');
                }
                if (this.stats.specs.excluded > 0) {
                    specsInfo.push(this.stats.specs.excluded + ' excluded');
                }
                if (specsInfo.length) {
                    this.print.str(this.style.yellow(' (' + specsInfo.join(', ') + ')'));
                }

                let exInfo;
                const totalExpects = this.stats.expects.total;
                this.print.line('Expects: ' + this.style.white(totalExpects));
                if (totalExpects === 0) {
                    exInfo = this.style.yellow(' (none executed)');
                } else {
                    const fc = this.stats.expects.failed;
                    exInfo = ' (' + fc + ' ' + utils.plural('failure', fc) + ')';
                    if (fc > 0) exInfo = this.style.red(exInfo);
                }

                this.print.str(exInfo);
            }

            this.print.line(this.style.gray('Finished in ' + this.stats.elapsed + ' ' + utils.plural('second', this.stats.elapsed)));
            this.print.newLine(2);
        } else {
            this.print.newLine();
        }

        this.stats.failedSuiteList.forEach(suite => {
            this._suiteFailureDetails(suite);
        });

        if (this.options.beep && this.stats.expects.falied > 0) {
            this.print.beep();
        }
    }

    jasmineStarted(summary) {
        this.stats.init(summary);

        this.print.newLine();

        if (summary.totalSpecsDefined === 0) {
            this.print.newLine();
            this.print.str(this.style.emoji.noSpecs + this.style.yellow('No specs defined!'));
            return;
        }

        this.print.str('Executing ' + summary.totalSpecsDefined + ' defined specs...');

        const isRandom = summary.order && summary.order.random;
        if (this.verbosity.summary && isRandom) {
            this.print.newLine();
            this.print.str(this.style.gray('Running in random order... (seed: ' + summary.order.seed + ')'));
        }
        this.print.newLine();

        if (this.verbosity.specs) {
            this.print.line(this.style.white(this.style.underline('Test Suites & Specs')) + this.style.white(':'));
            this.print.newLine();
        }
    }

    suiteStarted(suite) {
        this._depth++;
        this.stats.addSuite(suite);

        const isFirstSuite = this.stats.suites.total === 1;
        if (!isFirstSuite && this.verbosity.specs) {
            this.print.newLine();
        }

        if (this.verbosity.specs) {
            this._depth = this._depth || 0;
            const ind = this.options.listStyle === 'indent'
                ? utils.repeat(INDENT_CHAR, this._depth * INDENT_UNIT)
                : '';
            const title = this.style.cyan(this.stats.suites.total + '. ' + suite.description);
            this.print.line(ind + title);
        }

        this._isSuiteDone = false;
    }

    suiteDone(suite) {
        this.stats.updateSuite(suite);
        this._depth--;
        this._isSuiteDone = true;
    }

    specStarted(spec) {
        this.stats.addSpec(spec);
        if (this.verbosity.specs) {
            this.print.newLine(this._isSuiteDone ? 2 : 1);
        }

        // show the activity animation and current spec to be executed, if
        // enabled.
        if (this.options.activity) {
            const ind = this.verbosity.specs && this.options.listStyle === 'indent'
                ? utils.repeat(INDENT_CHAR, (this._depth + 1) * INDENT_UNIT)
                : '';
            this.print.spin(ind + this.style.gray(spec.description));
        }
    }

    specDone(spec) {
        this.stats.updateSpec(spec);
        if (this.options.activity) this.print.spinStop();

        if (this.verbosity.specs) {
            this._depth = this._depth || 0;
            let title = '';
            const ind = this.options.listStyle === 'indent'
                ? utils.repeat(INDENT_CHAR, (this._depth + 1) * INDENT_UNIT)
                : '';
            let timeStyle;

            switch (spec.status) {
                case 'pending':
                    title = this.style.yellow(this.style.symbol('warning') + ' ' + spec.description);
                    break;
                case 'disabled':
                case 'excluded':
                    // we don't print disableds and exludeds
                    if (!this.verbosity.disabled) {
                        // clear the new line printed on spec-start
                        // this.print.clearLine();
                        this.print.clearLine(2);
                        this.print.moveCursor(0, -1);
                        return;
                    }
                    title = this.style.gray(this.style.symbol('disabled') + ' ' + spec.description);
                    break;
                case 'failed': {
                    const fc = spec.failedExpectations.length;
                    const f = ' (' + fc + ' ' + utils.plural('failure', fc) + ')';
                    timeStyle = this.style.time(spec._time.num);
                    title = this.style.red(this.style.symbol('error') + ' ' + spec.description + f)
                        + timeStyle(' (' + spec._time.str + ')');
                    break;
                }
                case 'passed':
                    timeStyle = this.style.time(spec._time.num);
                    title = this.style.green(this.style.symbol('success') + ' ' + spec.description)
                        + timeStyle(' (' + spec._time.str + ')');
                    break;
                default:
                    // unknown status
                    break;
            }

            this.print.str(ind + title);
        }
    }

    jasmineDone(summary) {
        this.stats.done(summary);

        this.print.newLine();
        if (this.stats.specs.defined > 0) {
            if (this.verbosity.specs) this.print.newLine();
            this.print.str(this.style.gray('>> Done!'));
            this.print.newLine(2);
        }

        this._finalReport(summary);
    }
}

module.exports = JasmineConsoleReporter;


lib/activity.js
/**
 *  Activity Logger
 *  @author   Onur Yıldırım (onur@cutepilot.com)
 *  @license  MIT
 */
module.exports = (function () {
    'use strict';

    // ---------------------------
    //  CLASS: Activity
    // ---------------------------

    function defaultLogger() {
        process.stdout.write.apply(process.stdout, arguments);
    }

    // example:
    // var activity = new Activity(80);
    // activity.start('* please wait...');
    // asterisk will be replaced with rotating line animation (\ | / —) on
    // each interval. For ANSI codes, see:
    // http://academic.evergreen.edu/projects/biophysics/technotes/program/ansi_esc.htm
    function Activity(interval, log) {
        this._ticks = 0;
        this._interval = interval || 60;
        this.running = false;
        this._log = typeof log !== 'function'
            ? defaultLogger
            : log;
    }

    Activity.prototype.stop = function () {
        if (this._timer) {
            clearInterval(this._timer);
            this._timer = null;
        }
        // clear the full title
        if (this.running) {
            this._log('\x1B[u\r\x1B[K');
        }
        this.running = false;
        this._ticks = 0;
        this._row = null;
    };

    function _activityRun() {
        this._ticks += 1;
        const rotate = ['\\', '|', '/', '-'];
        const c = rotate[this._ticks % 4];
        const title = this.title ?
            this.title.replace(/\*/g, c)
            : c;
        // move cursor to last saved position, clear line and update activity
        // title.
        this._log('\x1B[u' + title);
    }

    Activity.prototype.start = function (title) {
        const $this = this;
        $this.stop();
        $this.title = title;
        $this.running = true;
        // save cursor position
        this._log('\x1B[s');
        $this._timer = setInterval(function () {
            _activityRun.apply($this);
        }, $this._interval);
    };

    return Activity;

})();


lib/utils.js
const path = require('path');

/**
 *  utils / helpers
 *  @author   Onur Yıldırım <onur@cutepilot.com>
 */
module.exports = (() => {

    const utils = {
        // windows returns 'win32' even on 64 bit but we still check for win64,
        // just in case...
        isWindows: process.platform === 'win32'
            || process.platform === 'win64'
    };

    utils.type = obj => ({}).toString.call(obj).match(/\s(\w+)/i)[1].toLowerCase();

    utils.isNum = value => typeof value === 'number';

    utils.log = (...args) => {
        process.stdout.write.apply(process.stdout, args);
    };

    utils.plural = (str, count) => count === 1 ? str : str + 's';

    utils.capitalize = string => string.slice(0, 1).toUpperCase() + string.slice(1);

    utils.repeat = (string, times) => new Array(times + 1).join(string);

    utils.indent = (str, times, indentChar) => {
        indentChar = indentChar || ' ';
        const newArr = [];
        const lines = (str || '').split('\n');
        for (let i = 0; i < lines.length; i++) {
            newArr.push(utils.repeat(indentChar, times) + lines[i]);
        }
        return newArr.join('\n');
    };

    let pStackLine = /[ \t]+at [^\r\n]+/g;
    utils.restack = (expectation, cleanLevel, style) => {
        if (!expectation) return '';

        let stack = String(expectation.stack || '');
        let message = typeof expectation.message === 'string'
            ? expectation.message
            : (stack.split(pStackLine)[0] || '');
        message = message.replace(/[\r\n]+$/, '');

        let stackLines = stack.match(pStackLine);

        if (cleanLevel > 0) {
            let cleanPath = cleanLevel >= 2
                ? 'node_modules'
                : path.join('node_modules', 'jasmine-core');
            cleanPath = path.sep + cleanPath + path.sep;
            // remove stack lines with jasmine-core path.
            stackLines = stackLines.filter(line => {
                const pathCheck = line.indexOf(cleanPath) === -1;
                const sepCheck = cleanLevel >= 3
                    ? line.indexOf(path.sep) > -1
                    : true;
                return (pathCheck && sepCheck);
            });
        }

        stackLines = stackLines.map(line => {
            // make the file paths CMD+clickable by removing the wrapping parenths.
            line = line.replace(/\(([^(]+?)\)/g, '$1'); // '( $1 )'
            if (style) {
                // add colors/styles to file name, line and column numbers
                line = line.replace(/([^:/\\ ]+):(\d+):(\d+)/, (m, $1, $2, $3) => {
                    return style.yellow($1) + ':' + style.white($2) + ':' + style.white($3);
                });
            }
            return line;
        });
        // add back the error message and rest of stack
        if (style) message = style.red(message);
        return message + '\n' + stackLines.join('\n');
    };

    utils.extend = (defaults, object) => {
        object = object || {};
        for (let key in defaults) {
            if (defaults.hasOwnProperty(key) && object[key] === undefined) {
                object[key] = defaults[key];
            }
        }
        return object;
    };

    // ---------------------------

    const defaultOpts = {
        colors: 1,              // 0 to 2
        cleanStack: 1,          // 0 to 3
        verbosity: {            // object or number (0 to 4)
            specs: true,
            failed: true,
            pending: true,
            disabled: true,
            summary: true
        },
        activity: false,        // bool or string ("dots")
        listStyle: 'indent',    // "indent"|"flat"
        timeUnit: 'ms',         // s|ms|ns
        timeThreshold: { ok: 500, warn: 1000, ouch: 3000 }, // number|object
        emoji: true,
        beep: true
    };

    // Normalizes and tries to get a numeric value. This is used for options
    // that support both Boolean and Number values.
    function _boolToNum(value, numTrue, numFalse, defaultVal) {
        if (typeof value === 'boolean') {
            return value ? numTrue : numFalse;
        }
        return typeof value === 'number' ? value : defaultVal;
    }

    function _mormalizeVerbosity(verbosity) {
        if (utils.type(verbosity) === 'object') {
            return Object.assign({}, defaultOpts.verbosity, verbosity);
        }
        const v = _boolToNum(verbosity, 4, 0, null);
        if (v === null) return Object.assign({}, defaultOpts.verbosity);
        return {
            failed: true, // always true
            disabled: v >= 4, // also list disabled specs and some additional info
            specs: v >= 3,
            pending: v >= 2,
            summary: v >= 1
        };
    }

    function _keep(n, min) {
        return typeof n === 'number' && n > (min || 0);
    }

    function _normalizeThreshold(threshold) {
        const t = Object.assign({}, threshold);
        if (!_keep(t.ok, 0)) delete t.ok;
        if (!_keep(t.warn, t.ok)) delete t.warn;
        if (!_keep(t.ouch, t.warn)) delete t.ouch;
        return t;
    }

    utils.getOptions = options => {
        let opts = options || {};
        opts.cleanStack = _boolToNum(opts.cleanStack, 1, 0, defaultOpts.cleanStack);
        opts.colors = _boolToNum(opts.colors, 1, 0, defaultOpts.colors);
        opts = Object.assign({}, defaultOpts, opts);
        opts.verbosity = _mormalizeVerbosity(opts.verbosity);
        const threshold = opts.timeThreshold;
        const t = utils.type(threshold);
        opts.timeThreshold = t === 'number'
            ? { ouch: threshold }
            : t === 'object'
                ? _normalizeThreshold(threshold)
                : Object.assign({}, defaultOpts.timeThreshold);
        return opts;
    };

    // ---------------------------

    return utils;

})();
